File: src/test/java/com/gildedrose/GildedRoseTest.java
package com.gildedrose;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

class GildedRoseTest {

    @Test
    void updateQulaityWithFooProduct() {
        //arrange
        Item[] items = new Item[]{new Item("foo", 0, 0)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("foo", app.items[0].name);
        Assertions.assertEquals(-1, app.items[0].sellIn);
        Assertions.assertEquals(0, app.items[0].quality);
    }

    @Test
    void updateQulaityWithAgedBrie() {
        //arrange
        Item[] items = new Item[]{new Item("Aged Brie", 4, 6)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("Aged Brie", app.items[0].name);
        Assertions.assertEquals(3, app.items[0].sellIn);
        Assertions.assertEquals(7, app.items[0].quality);
    }

    @Test
    void updateQulaityWithSulfuras() {
        //arrange
        Item[] items = new Item[]{new Item("Sulfuras, Hand of Ragnaros", 4, 6)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("Sulfuras, Hand of Ragnaros", app.items[0].name);
        Assertions.assertEquals(4, app.items[0].sellIn);
        Assertions.assertEquals(6, app.items[0].quality);
    }

    @Test
    @DisplayName("Test for wrong BackstagePasses. Acting like usual product")
    void updateQulaityWithWrongBackstagePasses() {
        //arrange
        Item[] items = new Item[]{new Item("Backstage passes", 8, 50)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("Backstage passes", app.items[0].name);
        Assertions.assertEquals(7, app.items[0].sellIn);
        Assertions.assertEquals(49, app.items[0].quality);
    }

    @Test
    @DisplayName("Test for corrrect BackstagePasses, when less than 10 days to sell")
    void updateQulaityWithCorrectBackstagePassesIncreaseByTwo() {
        //arrange
        Item[] items = new Item[]{new Item("Backstage passes to a TAFKAL80ETC concert", 8, 40)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("Backstage passes to a TAFKAL80ETC concert", app.items[0].name);
        Assertions.assertEquals(7, app.items[0].sellIn);
        Assertions.assertEquals(42, app.items[0].quality);
    }

    @Test
    @DisplayName("Test for corrrect BackstagePasses, when less than 5 days to sell")
    void updateQulaityWithBackstagePassesIncreseByThree() {
        //arrange
        Item[] items = new Item[]{new Item("Backstage passes to a TAFKAL80ETC concert", 3, 40)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("Backstage passes to a TAFKAL80ETC concert", app.items[0].name);
        Assertions.assertEquals(2, app.items[0].sellIn);
        Assertions.assertEquals(43, app.items[0].quality);
    }

    @Test
    @DisplayName("Test for corrrect BackstagePasses, when more than 10 days to sell")
    void updateQulaityWithBackstagePassesIncreseByOne() {
        //arrange
        Item[] items = new Item[]{new Item("Backstage passes to a TAFKAL80ETC concert", 20, 40)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("Backstage passes to a TAFKAL80ETC concert", app.items[0].name);
        Assertions.assertEquals(19, app.items[0].sellIn);
        Assertions.assertEquals(41, app.items[0].quality);
    }

    @Test
    void updateQulaityNoDaysLeft() {
        //arrange
        Item[] items = new Item[]{new Item("Some ITEM", 0, 40)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("Some ITEM", app.items[0].name);
        Assertions.assertEquals(-1, app.items[0].sellIn);
        Assertions.assertEquals(38, app.items[0].quality);
    }

    @Test
    void updateQulaityAgedBrieWithMaxQuality() {
        //arrange
        Item[] items = new Item[]{new Item("Aged Brie", 5, 50)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("Aged Brie", app.items[0].name);
        Assertions.assertEquals(4, app.items[0].sellIn);
        Assertions.assertEquals(50, app.items[0].quality);
    }

    @Test
    void updateQulaityAgedBrieWithNoDaysLeft() {
        //arrange
        Item[] items = new Item[]{new Item("Aged Brie", 0, 40)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("Aged Brie", app.items[0].name);
        Assertions.assertEquals(-1, app.items[0].sellIn);
        Assertions.assertEquals(42, app.items[0].quality);
    }

    @Test
    void updateQulaityAgedBrieWithLessThanFiveDays() {
        //arrange
        Item[] items = new Item[]{new Item("Aged Brie", 4, 40)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("Aged Brie", app.items[0].name);
        Assertions.assertEquals(3, app.items[0].sellIn);
        Assertions.assertEquals(41, app.items[0].quality);
    }

    @Test
    void updateQulaityBackstagePassesWithNoDays() {
        //arrange
        Item[] items = new Item[]{new Item("Backstage passes to a TAFKAL80ETC concert", 0, 40)};
        GildedRose app = new GildedRose(items);

        //act
        app.updateQuality();

        //assert
        Assertions.assertEquals("Backstage passes to a TAFKAL80ETC concert", app.items[0].name);
        Assertions.assertEquals(-1, app.items[0].sellIn);
        Assertions.assertEquals(0, app.items[0].quality);
    }


}

File: src/test/kotlin/com/gildedrose/GildedRoseTest.kt
package com.gildedrose

import org.junit.Assert.assertEquals
import org.junit.Test
import kotlin.math.exp

internal class GildedRoseTest {

    @Test
    fun `basic item should decrease quality by 1 when sellIn is positive`() {
        val items = arrayOf<BasicItem>(BasicItem("Basic Item", 1, 5))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Basic Item", 0, 4))
    }



    @Test
    fun `basic item should decrease quality by 2 when sellIn is lte 0`() {
        val items = arrayOf<BasicItem>(BasicItem("Basic Item", 0, 5))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Basic Item", -1, 3))
    }

    @Test
    fun `basic item should not change to negative quality when sellIn is positive`() {
        val items = arrayOf<BasicItem>(BasicItem("Basic Item", 1, 0))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Basic Item", 0, 0))
    }

    @Test
    fun `basic item should change to negative quality when sellIn is lte 0`() {
        val items = arrayOf<BasicItem>(BasicItem("Basic Item", 0, 1))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Basic Item", -1, 0))
    }

    @Test
    fun `basic item should not change quality when quality is min and sellIn is lte 0`() {
        val items = arrayOf<BasicItem>(BasicItem("Basic Item", 0, 0))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Basic Item", -1, 0))
    }

    @Test
    fun `aged brie item should increase quality by 1 when sellIn is gt 0`() {
        val items = arrayOf<BasicItem>(AgedBrieItem("Aged Brie", 1, 3))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Aged Brie", 0, 4))
    }

    @Test
    fun `aged brie item should increase quality by 2 when sellIn is lte 0`() {
        val items = arrayOf<BasicItem>(AgedBrieItem("Aged Brie", 0, 3))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Aged Brie", -1, 5))
    }

    @Test
    fun `aged brie item should not increase quality more than max when sellIn is gt 0`() {
        val items = arrayOf<BasicItem>(AgedBrieItem("Aged Brie", 1, 50))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Aged Brie", 0, 50))
    }

    @Test
    fun `aged brie item should not increase quality more than max when sellIn is lte 0`() {
        val items = arrayOf<BasicItem>(AgedBrieItem("Aged Brie", 0, 49))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Aged Brie", -1, 50))
    }

    @Test
    fun `aged brie item should not increase quality when sellIn is lte 0`() {
        val items = arrayOf<BasicItem>(AgedBrieItem("Aged Brie", 0, 50))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Aged Brie", -1, 50))
    }

    @Test
    fun `sulfuras item should not change quality nor sellIn`() {
        val items = arrayOf<BasicItem>(SulfurasItem("Sulfuras, Hand of Ragnaros", 0, 80))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Sulfuras, Hand of Ragnaros", 0, 80))
    }

    @Test
    fun `backstage passes should increase quality by 1 when sellIn gt 10`() {
        val items = arrayOf<BasicItem>(BackstagePassesItem("Backstage passes to a TAFKAL80ETC concert", 11, 5))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Backstage passes to a TAFKAL80ETC concert", 10, 6))
    }

    @Test
    fun `backstage passes should increase quality by 2 when sellIn lte 10 and gt 5`() {
        val items = arrayOf<BasicItem>(BackstagePassesItem("Backstage passes to a TAFKAL80ETC concert", 10, 5))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Backstage passes to a TAFKAL80ETC concert", 9, 7))
    }

    @Test
    fun `backstage passes should increase quality by 3 when sellIn lte 5 and gt 0`() {
        val items = arrayOf<BasicItem>(BackstagePassesItem("Backstage passes to a TAFKAL80ETC concert", 5, 5))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Backstage passes to a TAFKAL80ETC concert", 4, 8))
    }

    @Test
    fun `backstage passes should make quality 0 when sellIn is 0`() {
        val items = arrayOf<BasicItem>(BackstagePassesItem("Backstage passes to a TAFKAL80ETC concert", 0, 5))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Backstage passes to a TAFKAL80ETC concert", -1, 0))
    }

    @Test
    fun `backstage passes should not increase quality when quality is max and sellIn gt 10`() {
        val items = arrayOf<BasicItem>(BackstagePassesItem("Backstage passes to a TAFKAL80ETC concert", 11, 50))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Backstage passes to a TAFKAL80ETC concert", 10, 50))
    }

    @Test
    fun `backstage passes should not increase quality more than max and sellIn lte 10 and gt 5`() {
        val items = arrayOf<BasicItem>(BackstagePassesItem("Backstage passes to a TAFKAL80ETC concert", 10, 49))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Backstage passes to a TAFKAL80ETC concert", 9, 50))
    }

    @Test
    fun `backstage passes should not increase quality when quality is max and sellIn lte 10 and gt 5`() {
        val items = arrayOf<BasicItem>(BackstagePassesItem("Backstage passes to a TAFKAL80ETC concert", 10, 50))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Backstage passes to a TAFKAL80ETC concert", 9, 50))
    }

    @Test
    fun `backstage passes should not increase quality when quality is max and sellIn lte 5 and gt 0`() {
        val items = arrayOf<BasicItem>(BackstagePassesItem("Backstage passes to a TAFKAL80ETC concert", 5, 50))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Backstage passes to a TAFKAL80ETC concert", 4, 50))
    }

    @Test
    fun `backstage passes should not increase quality not more than max when sellIn lte 5 and gt 0`() {
        val items = arrayOf<BasicItem>(BackstagePassesItem("Backstage passes to a TAFKAL80ETC concert", 5, 49))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Backstage passes to a TAFKAL80ETC concert", 4, 50))
    }

    @Test
    fun `backstage passes should not increase quality not more than max when sellIn lte 5 and gt 0 and quality is max - 2`() {
        val items = arrayOf<BasicItem>(BackstagePassesItem("Backstage passes to a TAFKAL80ETC concert", 5, 48))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Backstage passes to a TAFKAL80ETC concert", 4, 50))
    }

    @Test
    fun `conjured items should decrease quality by 2 when sellIn is positive`() {
        val items = arrayOf<BasicItem>(ConjuredItem("Conjured Mana Cake", 5, 5))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Conjured Mana Cake", 4, 3))
    }

    @Test
    fun `conjured items should decrease quality by 4 when sellIn is lte 0`() {
        val items = arrayOf<BasicItem>(ConjuredItem("Conjured Mana Cake", 0, 5))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Conjured Mana Cake", -1, 1))
    }

    @Test
    fun `conjured items should not decrease quality more than min when sellIn is positive`() {
        val items = arrayOf<BasicItem>(ConjuredItem("Conjured Mana Cake", 5, 1))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Conjured Mana Cake", 4, 0))
    }

    @Test
    fun `conjured items should not decrease quality when quality is min and sellIn is positive`() {
        val items = arrayOf<BasicItem>(ConjuredItem("Conjured Mana Cake", 5, 0))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Conjured Mana Cake", 4, 0))
    }

    @Test
    fun `conjured items should not decrease quality more than min when sellIn is lte 0`() {
        val items = arrayOf<BasicItem>(ConjuredItem("Conjured Mana Cake", 0, 3))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Conjured Mana Cake", -1, 0))
    }

    @Test
    fun `conjured items should not decrease quality when quality is min and sellIn is lte 0`() {
        val items = arrayOf<BasicItem>(ConjuredItem("Conjured Mana Cake", 0, 0))
        val app = GildedRose(items)

        app.updateQuality()

        assertEqualsItems(app.items[0], Item("Conjured Mana Cake", -1, 0))
    }

    private fun assertEqualsItems(actual: Item, expected: Item) {
        assertEquals("Item name", expected.name, actual.name)
        assertEquals("Item sellIn", expected.sellIn, actual.sellIn)
        assertEquals("Item quality", expected.quality, actual.quality)
    }
}

